### PR TITLE
LibELF: Make orders in which we store/call things explicit

### PR DESCRIPTION
**Pull request by DanShaders**

### CODE REVIEW

**Comment by DanShaders** [comment]
Hmmm, I was convinced tests fail on master but it seems like I messed something up. Looks like `for_each_unfinished_dependency` actually does implicitly sort dependencies in topological order.

I still think this PR is an improvement though: at least, it adds spec links, makes the said orders completely explicit, and prevents useless conversions of DynamicLoader objects to their respective paths and back. I should probably just reframe it as a refactor, not doing this "more correctly".

**Comment by DanShaders** [comment]
@BertalanD I've split tests to a separate PR and made it clear in a commit title/description that this is a refactor.

File: Userland/Libraries/LibELF/DynamicLinker.cpp [hunk 1]
```diff
@@ -781,6 +762,9 @@ Examples of static-pie ELF objects are ELF packers, and the system dynamic loade
 #endif
     }
 
+    // Manually end DependencyOrdering lifetime to free DynamicLoader objects.
+    objects.~DependencyOrdering();
```
Comment (ADKaster) [inline Userland/Libraries/LibELF/DynamicLinker.cpp:766]: This is pretty unfortunate. could we store the objects in an optional and clear that optional instead? Calling destructors manually outside of data structure types makes me uncomfortable.

**Review by ADKaster — COMMENTED** [review]
One comment, but otherwise I like this refactor. It seems easier to follow than the original code.

**Review by ADKaster — APPROVED** [review]

**Pull request merged**